### ISSUE TITLE
Notification usage guidelines

### ISSUE THREAD

**Issue by isidorn** [microsoft/vscode#119463]
I think we should

- [ ] Revisit all internal usages of notifications and consider if they should be converted to modal dialogs if they require immediate user actions
- [ ] Once we figure out the best thing to do in VS Code core we should update our extension guidelines so extension can follow the patterns that we do in core

Ideally this exercise would move "important" to be modal dialogs, and thus we can start exercising a more aggressive hide strategy towards regular notifications.

fyi @devinvalenciano @meganrogge

**Comment by Tyriar** [comment]
The terminal workspace permissions dialogs should be modals imo, but that might get fixed when we adopt trusted workspaces https://github.com/microsoft/vscode/issues/119344

**Comment by isidorn** [comment]
@misolori is working on a nice flow chart to make it easy to distinguish if a usage should use a modal dialog or a notification.
Until he has that, here are some rules of thumb which I think can be helpful:

* User interaction immediately needed -> `modal`
* Feature can not continue without input -> `modal`
* Not directly user initiated -> `notification`

**Comment by isidorn** [comment]
Notifications which might no longer be needed:

- [ ] [FindWidget](https://github.com/microsoft/vscode/blob/main/src/vs/editor/contrib/find/findWidget.ts#L888) @rebornix 
- [ ] [Close Workspace action](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/browser/actions/workspaceActions.ts#L119) @sandy081 I think this action should be disabled when there is no workspace, so ideally this notification is not needed
- [ ] [Settings are automatically saved](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/preferences/browser/settingsEditor2.ts#L777) @sandy081 I think we have educated our users enough and I think the future is that setting do not have to be Applied. So I would think about removing this notification alltogether
- [ ] [Tests](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/testing/browser/testExplorerActions.ts#L277) shuold this be an `.error` or at least a `warn` @connor4312 
- [ ] [License](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/update/browser/update.ts#L192) @joaomoreno I think we can just remove this
- [ ] [Welcome](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/welcome/page/browser/welcomePage.ts#L627) @chrmarti this feels like it should use the progress api? And not notifications
- [ ] [Sync](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/services/userDataSync/browser/userDataSyncWorkbenchService.ts#L292) Sync turned on. @sandy081 do we really need to notify the user that sync is on, doesn't the activity bar badge update accordingly?

I have removed some notifications in debug and file land already and changed some from `info` to `warn` in `undoRedoService`.

**Comment by isidorn** [comment]
Here's a list of questionable Notifications which are used for user prompts:
- [ ] [MenuBar](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/browser/parts/titlebar/menubarControl.ts#L267) @sbatten we might be able to remove this since now the custom menu is on by default, correct?
- [ ] [LargeFileOptimisation](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/codeEditor/browser/largeFileOptimizations.ts#L47) @alexdima the flow here is -> we show notification with actions -> user chooses one -> we show another notification asking the user to reload the file. Could we improve this flow to just have one prompt notification which would just update the setting and automatically reload the file

**Comment by alexr00** [comment]
I removed the tasks prompt related to tasks v1 ✔.
However, I also added a new one to notify users that I upgraded their tasks from v1 to v2. This happens on startup, and few people will actually see it because of the low use of tasks v1:
![image](https://user-images.githubusercontent.com/38270282/115572863-93ade600-a2c0-11eb-873a-a496dd4fd406.png)

For the automatic tasks prompt, I expect we can move that to a setting and not show a prompt once workspace trust is turned on.

**Comment by isidorn** [comment]
@alexr00 great, thanks a lot! As for the the new notification it makes sense, however if the automatic upgrade just works I do not see a point in showing it. Though I guess it might not just work so that is the reason to show it. But in the long run (after 3 milestones) I would just remove it as well.
And yeah for the automatic task prompt - removing it next milestone would be great.

**Comment by alexr00** [comment]
I think we need to show something after the tasks upgrade. It will not be perfect, and the user will see that they have changes in their workspace when they go to commit. I want to make sure they know why that file changed.

**Comment by isidorn** [comment]
I have manually reviewed all internal usages of the `notificationService`. In the comment above I have listed all usages which I think should be discussed and maybe improved. For larger items I have created issues and linked them to this one.

I think we are actually not so bad in VS Code core, but I think we can still improve to be a **great role model** for our extension ecosystem.
I have assigned the owners to issues and checkbox items - thanks for your help!

**Comment by dbaeumer** [comment]
@isidorn regarding localization: I am not a fan of a modal dialog here. Best would be to have something during install that lets user choose to automatically switch to that locale.

**Comment by isidorn** [comment]
@dbaeumer I agree. Ideally it would work just like theme's where you install a theme and it is automatically applied. However the problem for localization is that it requires a restart right? If it does not require the restart we should just apply the language, but even a restart might not be that destructive and we might as well do it automatic?

**Comment by misolori** [comment]
## Notification decision tree

Here is our first attempt at creating a notification decision tree to help people figure out if they should use a notification or a modal dialogue. Would love to get feedback from everyone and to find any gaps that are missing.

<img width="1423" alt="CleanShot 2021-04-23 at 06 33 54" src="https://user-images.githubusercontent.com/35271042/115879099-2011f300-a3fe-11eb-846d-6a217740c802.png">

**Comment by sana-ajani** [comment]
I think there's another type of notification that falls in the "Feedback" category. Extensions show toast prompts asking users to take a survey to get user feedback or to get ratings (calculate NPS scores, ask people to support via $ or marketplace rating), but many extensions abuse this by sending surveys too soon or too often. 

I don't know if there's a best practice solution or recommendation here, because we do want extensions to get user feedback early on in a user's experience. From a user's perspective, I don't mind seeing these requests for feedback, but I get inundated when there's 2-3 of these toasts within the same session. One thing to note, if extensions are bundled together, either in a pack or like the Python + Jupyter extension, I think they should be sending surveys as a bundle, instead of for each extension in the bundle.

cc @digitarald @devinvalenciano

**Comment by misolori** [comment]
@sana-ajani what is our logic for sending NPS feedback? After xx number of actions? Time?

**Comment by sana-ajani** [comment]
This is where it's a bit tricky with the survey notifications. Our [logic for NPS](https://github.com/microsoft/vscode/blob/94c9ea46838a9a619aeafb7e8afd1170c967bb55/src/vs/workbench/contrib/surveys/browser/nps.contribution.ts) is okay, the user has a 15% chance to see prompt after 10 sessions. We use this logic for our core NPS survey, and for the Java, JS/TS, and C++ extension surveys. We also have the [new customer effort survey](https://github.com/microsoft/vscode/blob/94c9ea46838a9a619aeafb7e8afd1170c967bb55/src/vs/workbench/contrib/surveys/browser/ces.contribution.ts) for beginners + people who might churn before making it to 10 sessions which is time-based, they see the prompt after 1 hour of use, within the first 24 hours of installation. (@digitarald is still experimenting with this timing/logic). These are both one and done survey prompts, the user isn't asked again after they click on the survey.

But other extensions (i.e. Python/Jupyter, Docker, .NET) have their own logic to send surveys, some time-based, some action-based, and some might repeat or ask again. We also do a handful of [one-off experimental survey prompts](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/experiments/browser/experimentalPrompt.ts) for extensions that don't have a long-running survey mechanism.

I think this is a good opportunity to standardize on best practices for survey prompts; perhaps a rotating number of X surveys from our core and first party extensions, or limiting the number of survey prompts a user can see within X weeks or X sessions?

**Comment by digitarald** [comment]
> These are both one and done survey prompts, the user isn't asked again after they click on the survey.

Clarification on one and done: For most users the survey notifications will only show once. As they are sticky they require a choice. Edge cases are restarting/closing VS Code while the notification is visible, which causes them to trigger again.

> I think this is a good opportunity to standardize on best practices for survey prompts; perhaps a rotating number of X surveys from our core and first party extensions, or limiting the number of survey prompts a user can see within X weeks or X sessions?

I was wondering in the past if the current surveys could cause problems and leaned towards no, but a deeper analysis might help to clarify the assumptions:

- Usage-based timing (language files, etc) should space multiple surveys out across different sessions/days
- Random sampling means users don't get several survey prompts

> But other extensions (i.e. Python/Jupyter, Docker, .NET) have their own logic to send surveys, some time-based, some action-based, and some might repeat or ask again.

Another idea is to fold language surveys into VS Code's existing surveys by piping extension/language signals into the survey that unlock additional language questions (tracked in https://github.com/microsoft/vscode/issues/121883 for new-user survey). Python only diverged from the existing survey system because they wanted to ask earlier, so this could bring them back into the fold.

**Comment by dbaeumer** [comment]
@isidorn I forgot about the reload which makes things more tricky. May be a modal dialog is the best at the end of the day because of that reload. 

I think we can keep the discussion here. No need for a separate issue.

**Comment by isidorn** [comment]
@dbaeumer great, thanks! Let me know how it goes with the modal dialog approach.

As for the surveys I discussed a bit with @digitarald and first of all it would be cool if we have just one infrastructure in VS Code core for surveys - currently there are 3. Once we have that cleaned up we should have one logic when to show surveys (so they are not on startup). And then we can write a recommendation which we can share with all our extensions regarding when and how to show surveys.

**Comment by misolori** [comment]
Here is the latest notification decision tree w/ feedback from others:

![image](https://user-images.githubusercontent.com/35271042/116321585-3063fe00-a76f-11eb-9c19-0f2928d58087.png)

**Comment by sana-ajani** [comment]
Installed the LiveServer extension and got hit with this. Show details takes me to the extension repo. So this is another type of notification ("announcement") to include in extension guidelines:
![image](https://user-images.githubusercontent.com/9721254/116327343-957e1a80-a794-11eb-97f1-49ca5b629bce.png)

**Comment by digitarald** [comment]
Agreed. Good match for the *Explore integration of release notes and extension changelogs* that's in the iteration plan and already has mock ups for the Welcome page.

**Comment by alexdima** [comment]
I suggest that we also offer some kind of API which says "show notification if no other notifications are currently shown".

For example, here's what I got today. The survey notification should really not show at the same time when 2 other notifications are shown.
<img width="491" alt="Screen Shot 2021-04-29 at 10 28 24" src="https://user-images.githubusercontent.com/5047891/116522616-ae421980-a8d5-11eb-810c-e71b5c22ca46.png">

**Comment by isidorn** [comment]
@alexdima good point, we already have discussions bout this here, I outline the same point in this comment https://github.com/microsoft/vscode/issues/121802#issuecomment-824857419
Though this holds particularly true for survey notifications

**Comment by JacksonKearl** [comment]
Re using status bar progress for low priority notifications, I've been using that to show the download of the search index in remote hub and I've gotten feedback from a few different people (@kieferrm , @lszomoru , maybe others) that they did not see it at all. I think there's a general concern that people aren't going to look down to the status bar for progress notifications while they are focused on elements elsewhere in the editor.

**Comment by isidorn** [comment]
@JacksonKearl yeah I think that's the point, if they are low priority it is fine if some users do not notice it.
Things like download progress of the search index sounds to me like a perfect candidate for the status bar. Unless it was triggered by the explicit user action in the search view, than on top of this showing a progress spin in the search view makes sense. Though you understand this flow better than me so feel free to ping me and we can connect on this in a short meeting.

**Comment by misolori** [comment]
I've pushed changes to the vscode website and added the decision tree from above. If anyone has any additional changes/feedback, please add them here. Thanks!

**Comment by isidorn** [comment]
@misolori thanks for updating the guidelines 👏 
@devinvalenciano can we now share [these guidelines](https://code.visualstudio.com/api/references/extension-guidelines#notifications) with some of our partner extensions?

**Comment by yunusefendi52** [comment]
I just updated vscode to latest version, the formatter dialog now prevents me from typing because now it's blocking me, from the diagram above which desicion is this taken from? In my opinion the formatter dialog should not be blocking, I know it's user triggered but at the same time it's triggered while typing which should not be blocked

Note: Previous version if there is no formatter, it's showing prompt instead of dialog (which IMO is the correct one)

**Comment by isidorn** [comment]
We are not blindly following the diagram, it is just a recommendation on the best flow.
As for the formatter if you think this interrupts the flow please file a new issue with repro steps and ping me on it @isidorn and we can discuss it further. Thanks
